### ISSUE TITLE
Sticky headers/footers violation of SC 2.4.7?

### ISSUE THREAD

**Issue by JAWS-test** [w3c/wcag#952]
If the focus on the focused element is clearly visible, but the element is not in the visible area: is this  a violation of SC 2.4.7: "Focus Visible"?

The problem occurs frequently, e.g. with
* header or footer marked with `position:fixed`
* Pop-up conceals focused elements
* Custom made widget with aria-activedescendant: focused element is not scrolled into the visible area because the focus is on the parent element.

If this is a violation of 2.4.7, can I add an appropriate addition to the Understanding?
If not: can we add it to [2.4.11](https://github.com/w3c/wcag/pull/936) (WCAG 2.2)?

**Comment by mraccess77** [comment]
If and element is behind something else and you can't visually see the focus indicator then it's not clearly visible and in my book fails 2.4.7.  Not sure how it can be visible if it's not actually on-screen.    What is tricky is that this needs to be tested in different responsive breakpoints because sticky headers and footers can be different sizes in RWD and in some situations this may be broken and other it is not.  Also the situation can arise where part of the focus indicator is visible but not all of it.

**Comment by JAWS-test** [comment]
> arise where part of the focus indicator is visible but not all of it

As long as the focused element is fully visible, it should not be a violation of SC 2.4.7 if the focus indicator is not fully visible, should it?

**Comment by jake-abma** [comment]
yes

**Comment by alastc** [comment]
Making that a blanket statement (sticky footers block you from seeing the focus indicator is a fail) seems, um, over-broad.

I'd agree that if a sticky-footer blocks the view of links (and therefore focus styles) under all conditions, then yes, that is a 2.4.7 fail. It is equivalent to being off-screen. That would probably only happen for links at the bottom of the page though?

If the sticky footer is dismissable, presumably that would not fail?

How about if you can tab forwards, and reverse and see the focus indicator?

I wonder, is there a good solution to this yet? I have a feeling someone was talking about a CSS method where you could have a sticky footer/header which doesn't get in the way of tabbing through links... but I can't remember that off hand.

**Comment by JAWS-test** [comment]
@alastc 
I don't know any example where fixed headers and footers permanently hide the interactive elements. As a keyboard user, I usually have the following two ways to work around the problem:
* scrolling with the arrow keys (not possible if focus is on a selection list or radio button group)
* tab to the next element with the Tab key and back (not possible if it is the first or last element of the page; also not possible if there are fixed headers and footers and the element is hidden by the header when navigating back)

I would still consider it a problem if the focused element is not immediately visible when it gets the focus.

The problem occurs not only with footers, but often also with headers, e.g. if the page has a two-column layout: first you navigate down in the left navigation bar and then you jump to the first interactive element in the main area of the page, which is then hidden by the header.

CSS makes it very easy to create fixed headers and footers without obscuring interactive elements. The "trick" is to make only the visible area scrollable (without header and footer) and not the whole page. I can create a Technique for it.

**Comment by alastc** [comment]
> I would still consider it a problem if the focused element is not immediately visible when it gets the focus.

I agree it is a problem, the question is whether it is a reasonably solvable problem. (I think it's too late to ban sticky headers/footers.)

> The "trick" is to make only the visible area scrollable

Yes, I've thought about that approach, but it seems very hard to do well in a responsive way that also passes other criteria. For example, with different screen sizes / ratios does the text in the header/footer get cut off? Do within-page links work? Does scrolling work properly (with keyboard etc) if the main area is actually a sub-section of body?

The type of behaviour that I think would solve this is to have sticky elements that can expand/reflow and impact the main area. Then make the `main` scroll as though it is the `body` element (like with fixed position + overflow), so it is the page-level scrolling element.

That would require the user-agents to update though, I don't think it's possible at the moment (although I'd love to be wrong about that!).

**Comment by JAWS-test** [comment]
> I think it's too late to ban sticky headers/footers.

Too late for 2.4.7 or for 2.4.11?

> Yes, I've thought about that approach, but it seems very hard to do well in a responsive way that also passes other criteria. For example, with different screen sizes / ratios does the text in the header/footer get cut off? Do within-page links work? Does scrolling work properly (with keyboard etc) if the main area is actually a sub-section of body?

I haven't tested all the problems you mentioned intensively, but I wonder to what extent this is relevant:
* The problems also occur with the classic fixed headers and footers or
* the problems do not occur (in my opinion this is true for most of them) or
* the problems can easily be solved (media queries) or
* the problems can't be solved, but that wouldn't be a reason to accept a lack of focus visibility because of it (because fixed headers and footers could be avoided or the scroll position can be controlled by JS)

**Comment by alastc** [comment]
We would need to be sure the problems _can_ be solved, sticky elements are a common pattern, there needs to be a solution. A good next step would be to find a good example that doesn't cause tabbing issues.

**Comment by JAWS-test** [comment]
> A good next step would be to find a good example that doesn't cause tabbing issues.

I think [my example](https://codepen.io/jaws-test/pen/yLLWBmv) could be used for SC 2.4.7 and SC 1.4.10 because it solves the known problems for fixed header and footer that violate these two SCs

**Comment by alastc** [comment]
So we discussed this on the call on Tuesday, and there are two things here:

1. Does this scenario fail the letter of the 2.4.7 SC text? We were leaning towards no on this question, as the indicator is visible and "a mode of operation" would include being able to scroll. However, it wasn't unanimous.
2. Is there a reasonable solution?

It would be useful to create a technique for this, if the answer to 1 is yes, it would be a sufficient technique. If no, it would be advisory (like the media-queries for sticky headers one).

I'm afraid a simple test case doesn't really tell us if that type of approach works in practice. Real sites have more complex layouts, and especially when you include text-spacing (or internationalization), the likelyhood of things wrapping goes up enormously. 

In the test case, if the height of the box increases then the header has it's own scrolling area which is not a good thing. We've yet to find a 'good' example in the wild, and I have [asked around](https://twitter.com/alastc/status/1260347064002756610).

The [scroll-padding property](https://drafts.csswg.org/css-scroll-snap-1/#propdef-scroll-padding) is promising, in Chromium based browsers that can be used in a similar way to fixed-position. The downside of the header/footer increasing in height is less negative, you'd get some overlap only for the focus indicator, not breaking the layout. 

That aspect of scroll-snap doesn't work in Firefox, but I've raised an [issue on CSS scroll snap 1](https://github.com/w3c/csswg-drafts/issues/5066) to see if a spec update is possible, and that would make a bug on firefox more likely to work. (If it's that way around, I might need to try FF first.)

We'll tackle the question of whether this scenario should fail 2.4.7 again next week.

**Comment by JAWS-test** [comment]
Perhaps it can be considered in the discussion that
* I have given more examples in the initial comment where the focused element is not visible (e.g. when using aria-activedescendant) and in these cases the focused element cannot be scrolled into the visible area with the keyboard
* with fixed header and footer scrolling into the visible area is not always possible (see https://github.com/w3c/wcag/issues/952#issuecomment-555364931)

**Comment by alastc** [comment]
For things like a custom drop-down, shouldn't the focus move when you are selecting items? From a quick google for examples of aria-activedescendant, [some don't open](http://oaa-accessibility.org/example/11/), [some are problematic](https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-actions-active-descendant.html) from this point of view. However, what would a solution there be?

For items permanently hidden by sticky elements, I think everyone is agreeing that would be a failure (although I might bring that up under focus-order rather than focus visible).

**Comment by mraccess77** [comment]
I find it less likely that a focus indicator would be visible but the element not visible.  If the element is focused but there is no focus indicator because it's behind something and the user must scroll it might be conformant -- however, it's been my experience that this issue happens at the bottom of pages and often there is no way to scroll to make it visible again -- or scrolling to make it visible causes something else to go off screen.  So in my opinion:
1. If you can't scroll it into view without changing the zoom level to something less than 200% it fails.
2. If scrolling the focused element into view causes any needed functionality to disappear it would fail - example sticky headers with buttons that relate to functionality in the visible page.
3. If you can't scroll it into view it fails.

**Comment by alastc** [comment]
> it's been my experience that this issue happens at the bottom of pages and often there is no way to scroll to make it visible again

Presumably that is when you have increased the zoom level beyond the default? I think that would be an extension of the page variations concept so I'd consider that covered, so I'd agree with 1 & 3.

I'm not sure about 2, if the needed button is on the sticky header, how does that disappear?

**Comment by JAWS-test** [comment]
> However, what would a solution there be?

https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-actions-active-descendant.html is a good example. If the button is located at the bottom of the browser window, the list is opened with the keyboard and then navigated through the list entries with the arrow keys, the focused list entries are not scrolled into the visible area and are therefore not recognizable. Scrolling is also difficult because arrow key operation does not trigger scrolling. The solution would be to avoid aria-activedescendant or to ensure via JS that the focused element is displayed in the visible area.

**Comment by alastc** [comment]
> The solution would be to avoid aria-activedescendant or to ensure via JS that the focused element is displayed in the visible area.

That doesn't seem like a viable solution. Regardless of ARIA usage, a custom element with that that form of focus (within itself) is something people use. Failing it depending on how much the user has scrolled is not a good testing approach.

I've found people who have tried using JS to keep the focus in view, but no-one has been able to provide an example of a live real site. It seems hacky and not a robust method, although I'd love to be corrected on that.

**Comment by alastc** [comment]
Hi, this was discussed in the group and we didn't get to a consensus position on whether temporarily obscuring the focus indicator would fail 2.4.7. However, we will be investigating it as part of future work, possibly as part of the Focus Visible Enhanced SC, or in a a subsequent version.

**Comment by DavidMacDonald** [comment]
After our discussion last week about 2.4.7 I took the liberty of contacting the two previous chairs of WCAG 2.0 to get their recollections of 2.4.7 and whether it applies to layered content where focus is temporarily obscured behind a layer. Here is my question to them with their answers.

===David MacDonald's question===

There is a very animated WCAG discussion on the working group about SC 2.4.7 focus visible, 

Any keyboard operable user interface has a mode of operation where the keyboard focus indicator is visible.

The group has no consensus about whether it was intended to cover the scenario where the user tabs to a component which is behind a sticky header/footer etc, and is temporarily obscured, but if the user scrolls (or presses the spacebar) the component becomes visible with its focus ring visible.
My sense was that if the component can be shown with a simple action like scrolling and the focus ring is correctly implemented, then the SC would be met... but it's quite a dynamic discussion.

I'm wondering about your thoughts about that which would help provide context. 

=== Loretta Guarino Reid (PhD Computer Science), WCAG 2.0 chair and accessibility at Google ===

Nice to hear from you! I hope all is well with you and yours.

This is a very different question from the sorts of things I remember us discussing. The use of layers of UI elements and regions was not common then, and we were more focused on making sure you could find the focused element in the two-dimensional area of the window.

I'm not sure what I think the answer should be, either. For someone with low vision, having the focus highlight (and focused element) completely obscured poses real challenges. But how could such situations be avoided in the 2.5 dimensional world we find ourselves in? Would there need to be some user agent support for rearranging the top layer to uncover the focused item? How would the user know what to do? Yet, I can't see the web going back from these design patterns.

No wonder this has lead to extended discussion!

I'm afraid I am not being much help, but this is certainly an interesting problem to think about.

====== Gregg Vanderheiden (PhD Engineering, former WCAG 1 and WCAG 2 chair)  ====
Here is another one for the history books.  (That is — the place where they archive all the history and lost information on the guidelines.    They created one a while ago.  Do you know where it is?)

MY MEMORY
What we WANTED to say was that “the Keyboard focus indicator was highly visible”.   But that is not testable.    So we went with “visible” and used that as an anchor to provide advisory notes that the focus indicator should be highly visible — that is — visible from 10 times the normal viewing distance or more.     

So as it is - - it is pretty hard to not meet it - unless you do something crazy and don’t make the keyboard focus indicator visible at all — in which case it is not a keyboard focus indicator….   That is - you don’t have one at all. 

===David's Summary ====

- It was not the WCAG Working Group's intention when creating SC 2.4.7 to address layered content where a properly coded focus ring is temporarily obscured but can be shown by scrolling. 
- SC 2.4.7 wording doesn't apply to layered content, 
- There is no historical precedence to interpret it that way.
- There may be good reason to look at this as a new requirement for WCAG.NEXT, but layering is here to stay and this goes well beyond the question of sticky headers and footers, and is likely something that should be addressed at the browser level

The institutional memory homepage is here https://www.w3.org/WAI/GL/wiki/WCAG20/Institutional_Memory There is no content in there for 2.4.7.

I would add that in 20 years this is the first time it's been asked to us that I'm aware of. The request was from a power tester and not an end user. My guess for the reason we've not heard about it from end users, is the way the sighted keyboard users interact with content. I cover that in this blog entry.
https://www.davidmacd.com/blog/sighted-keyboard-only-user.html

**Comment by alastc** [comment]
In today's [meeting](https://www.w3.org/2020/06/02-ag-minutes.html#item05) (and via survey) we did come to consensus that, whilst it is an issue, a sticky footer/header which temporarily obscures the focus indicator does **_not_** trigger a failure of 2.4.7.

The next step is to try and address it in the new focus-visible SC. @mbgower took an action to create a PR for that update, hopefully he can branch from the [FV updates branch](https://github.com/w3c/wcag/tree/wcag22-focus-visible-enh-updates), and tag this issue...
